WWST-2441, WWST-2448, WWST-2449, WWST-6168, WWST-6171, WWST-6172 - Custom DTH for Qubino Dimmers (DIN, Flush, Flush 0-10V)

KKlimczukS opened the pull request:
WWST-2441, WWST-2448, WWST-2449, WWST-6168, WWST-6171, WWST-6172
Custom DTH for Qubino Dimmers (DIN, Flush, Flush 0-10V)

- support for Qubino Temperature Sensor ZMNHEA1 (as a child device) - this device doesn't work independently, it must be connected to another Qubino device
- device settings are exposed (those one that are common for all three dimmers)

@greens @dkirker @tpmanley Please, could You take a look at the code ?

cc :@MWierzbinskaS @ZWozniakS  @PKacprowiczS @BJanuszS @MGoralczykS

KKlimczukS committed:
New DTH for Qubino Flush Dimmer 0-10V - ZMNHVD

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,731 @@
+	commands << zwave.associationV1.associationSet(groupingIdentifier:1, nodeId:[zwaveHubNodeId]).format()
+	commands << zwave.associationV1.associationSet(groupingIdentifier:2, nodeId:[zwaveHubNodeId]).format()
+	commands << zwave.associationV1.associationSet(groupingIdentifier:3, nodeId:[zwaveHubNodeId]).format()
+	commands << zwave.associationV1.associationSet(groupingIdentifier:4, nodeId:[zwaveHubNodeId]).format()
```
can you explain what's going on here?

KKlimczukS replied:
@greens 
There probably should be **nodeId:[zwaveHubNodeId]** added to associationRemove calls.
Please correct me if i'm wrong -> the intention was to remove the hub from those association groups and then add it again - to avoid doubling hubNodeId in device's association table (in a situation where the device will be unpaired and then paired with the hub again).
If a device deals with it on the protocol level then that associationRemove calls are redundant here and they will be removed from the code.

greens replied:
Removing a device from the network should clear its association table.

I mostly just wanted a comment in the code to explain this non-standard code, but it sounds like the removes aren't necessary.

KKlimczukS committed:
Add associations with groups:5,6

KKlimczukS committed:
Shortened list of device settings

KKlimczukS committed:
Reversed synchronization of device settings

KKlimczukS committed:
Add a child illuminance sensor

KKlimczukS committed:
Handle events coming from different types of inputs. Code refactoring.

KKlimczukS committed:
Shortened list of device settings (removed settings of minimum/maximum dimming value)

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+
+def updated() {
+// Device-Watch simply pings if no device events received for 32min(checkInterval)
+	sendEvent(name: "checkInterval", value: 2 * 15 * 60 + 2 * 60, displayed: false, data: [protocol: "zwave", hubHardwareId: device.hub.hardwareID, offlinePingable: "1"])
```
This device shouldn't be offline pingable. It's intended only for life safety devices.

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+	commands << zwave.associationV1.associationSet(groupingIdentifier:4, nodeId:[zwaveHubNodeId]).format()
+	commands << zwave.associationV1.associationSet(groupingIdentifier:5, nodeId:[zwaveHubNodeId]).format()
+	commands << zwave.associationV1.associationSet(groupingIdentifier:6, nodeId:[zwaveHubNodeId]).format()
+	commands << zwave.multiChannelV3.multiChannelEndPointGet().format()
```
it sounds like, from your description, that some of these should only be set once it's detected that the child device is connected

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+    }
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelSet cmd, ep = null) {
```
when does the device send `set`s?

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+        def childDthName		= childDeviceSpecifics[map.name].childDthName
+        def childDthNamespace 	= childDeviceSpecifics[map.name].childDthNamespace
+
+        createChildDevice(childDthNamespace.toString(), childDthName.toString(), childDni.toString(), childComponentLabel.toString())
```
all of these should already be strings, right?

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+    ]
+}
+
+def createChildDevice(childDthNamespace, childDthName, childDni, childComponentLabel) {
```
the way you're using this, `childDni` is already the parent DNI + the endpoint number

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+def zwaveEvent(physicalgraph.zwave.commands.switchmultilevelv3.SwitchMultilevelStopLevelChange cmd) {
+	log.debug "SwitchMultilevelStopLevelChange: ${cmd}"
+	[:]
+}
```
any reason you're logging these separately?

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+	updateDataValue("MSR", msr)
+	updateDataValue("manufacturer", cmd.manufacturerName)
+	createEvent([descriptionText: "$device.displayName MSR: $msr", isStateChange: false])
+}
```
we don't need to do this anymore

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+
+	commands << zwave.basicV1.basicGet().format()
+	commands << zwave.switchMultilevelV3.switchMultilevelGet().format()
+	commands << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 1, scale: 0).format()
```
What is this get for? Aren't all the sensors on endpoints? Wouldn't this need to be encapsulated or only sent conditionally?

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+	def commands = []
+
+	commands << zwave.basicV1.basicGet().format()
+	commands << zwave.switchMultilevelV3.switchMultilevelGet().format()
```
if you're doing a multichannel get you shouldn't need to do a basic get and vice versa

greens commented on devicetypes/qubino/qubino-flush-dimmer-0-10V.src/qubino-flush-dimmer-0-10V.groovy:
```diff
@@ -0,0 +1,671 @@
+			"Valid only if input is set as mono-stable (push button)."
+	],
+	[
+		name: "Unsecure / Secure Inclusion", key: "unsecure/SecureInclusion", type: "boolean",
```
not sure how this would even be used, since you'd have to remove the device and re-add, and I don't think configuration persists between remove/installs.

greens commented on devicetypes/smartthings/child-illuminance-sensor.src/child-illuminance-sensor.groovy:
```diff
@@ -0,0 +1,55 @@
+ *
+ */
+metadata {
+	definition(name: "Child Illuminance Sensor", namespace: "smartthings", author: "SmartThings", ocfDeviceType: "oic.d.sensor") {
```
i think this needs a ui-metadata reference

greens commented:
It's not exactly clear because of the parameters, but is this device a switch that can have a temperature or illuminance sensor added to it, a switch that can be changed to act not as a switch but as a temperature or illuminance sensor, or something else?

Is it possible for this device to be reporting both temperature and illuminance at the same time?

KKlimczukS committed:
WWST-2441, WWST-2448, WWST-2449, WWST-6168, WWST-6171, WWST-6172 - Custom DTH for Qubino Dimmers (DIN, Flush, Flush 0-10V)

KKlimczukS commented:
@greens @dkirker 
I've refactored the code so all three qubino dimmers can be supported by one DHT.

@MWierzbinskaS @BJanuszS - Please, let me know if it works correctly with Your Qubino Dimmers.

KKlimczukS committed:
removed child illuminance sensor

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+			)
+
+			switch(it.type) {
+				case "boolRange":
```
just inline all of this.

KKlimczukS replied:
Fixed. Please let me know if it looks better now.

greens reviewed (changes requested):
There are no "boolRange" preferences, but a lot of boilerplate code for dealing with them. It should be removed. I'd also like if you could take it out of the walli DTHs since the same is true for them. In general, I'd like the boilerplate to be pared down to the actually represented types of preferences.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+
+private notNullCheck(value) {
+	return value != null
+}
```
I don't see the value of breaking this out as a function

KKlimczukS replied:
ok, fixed.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+def zwaveEvent(physicalgraph.zwave.commands.multichannelv3.MultiChannelEndPointReport cmd, ep = null) {
+	log.info "MultiChannelEndPointReport: ${cmd}"
+	response([
+			refresh()
```
why does this message trigger a refresh?

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+
+	if(input1SwitchType == INPUT_TYPE_POTENTIOMETER) {
+		log.debug "BasicSet: ${cmd} / INPUT_TYPE_POTENTfIOMETER"
+		[response(zwave.switchMultilevelV3.switchMultilevelGet().format())]
```
you don't need to call `.format()` here, `response()` accepts zwave command objects.

Also why is it wrapped in brackets?

KKlimczukS replied:
yes, wrapping was unnecessary. Fixed.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+
+private dimmerEvents(physicalgraph.zwave.Command cmd, ep = null) {
+	createDimmerEvents(cmd.value)
+}
```
why is this a separate function?

KKlimczukS replied:
Yeah, that didn't make sense. Fixed.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		log.debug "handleChildEvent / creating a child device"
+		def childComponentLabel	= "Qubino Temperature Sensor"
+		def childDthName		= "Child Temperature Sensor"
+		def childDthNamespace 	= "qubino"
```
if these are always the same, just inline them. You can create constants for them if you want to avoid magic strings.

KKlimczukS replied:
Good point. Fixed.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		def childDthNamespace 	= "qubino"
+
+		createChildDevice(childDthNamespace, childDthName, childDni, childComponentLabel)
+		childDevice = childDevices.find { it.deviceNetworkId == childDni }
```
`createChildDevice` will return the child device if successful. You don't need to go and get it.

KKlimczukS replied:
Good point. Fixed, it should work more efficient now.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+	], 3000)
+}
+
+def setLevel(value, duration) {
```
this and the durationless `setLevel` can easily be combined.

KKlimczukS replied:
Fixed (the code was refactored).

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		def childDevice = childDevices.find { it.deviceNetworkId == childDni }
+
+		if (childDevice != null) {
+			sendHubCommand(zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 1, scale: 0).format())
```
you don't need to `format` this to pass it to `sendHubCommand`

greens replied:
also shouldn't this get encapsulated?

KKlimczukS replied:
Yup, it should. It's fixed now.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+	def commands = []
+	parameterMap.each {
+		state.currentPreferencesState."$it.key".status = "reverseSyncPending"
+		commands += encap(zwave.configurationV2.configurationGet(parameterNumber: it.parameterNumber))
```
you could just collect all of the commands and then call `encapCommands` on the whole list

KKlimczukS replied:
Fixed (I refactored the code, it should look better now).

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		name: "Dimming duration", key: "dimmingDuration", type: "range",
+		parameterNumber: 68, size: 1, defaultValue: 0,
+		range: "0..127",
+		description: "This parameter is used with association group 3. The Duration field MUST specify the time that the transition should take from the current value to the new target value. " +
```
is it useful to tell users about association group 3?

KKlimczukS replied:
Nah, i removed that sentence.

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		name: "Dimming time when key pressed", key: "dimmingTimeWhenKeyPressed", type: "range",
+		parameterNumber: 66, size: 2, defaultValue: 3,
+			range: "1..255",
+			description: "Time of moving the dimmer between min. and max dimming values by continues hold of push button I1 or associated device. " +
```
"The time it takes for the dimmer to transition between min and max brightness when push button I1 or other associated device is held continuously"

KKlimczukS replied:
Fixed

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		name: "Dimming time (soft on/off)", key: "dimmingTime(SoftOn/Off)", type: "range",
+		parameterNumber: 65, size: 2, defaultValue: 100,
+		range: "50..255",
+		description: "Set value means time of moving the dimmer between min. and max. dimming values by short press of push button or controlled through UI (BasicSet). " +
```
"The time it takes for the dimmer to transition between min and max brightness after a short press of the button or when controlled through the UI"

KKlimczukS replied:
Fixed

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		parameterNumber: 30, size: 1, defaultValue: 0,
+		optionInactive: 0, inactiveDescription: "Default value - dimmer module saves its state before power failure (it returns to the last position saved before a power failure)",
+		optionActive: 1, activeDescription: " Flush Dimmer 0-10V module does not save the state after a power failure, it returns to off position",
+		description: "Based on the parameter settings the stores/does not store the last value of the output after power failure. "
```
"Set whether the device stores or does not store the last output level in the event of a power outage."

KKlimczukS replied:
Fixed

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		parameterNumber: 21, size: 1, defaultValue: 0,
+		optionInactive: 0, inactiveDescription: "Default value - Double click disabled",
+		optionActive: 1, activeDescription: "Double click enabled",
+		description: "If Double click function is enabled, a fast double click on the push button will set dimming power at maximum dimming value. " +
```
"If enabled, a fast double-click on the push button will set the dimming level to its max."

KKlimczukS replied:
Fixed

greens commented on devicetypes/qubino/qubino-dimmer.src/qubino-dimmer.groovy:
```diff
@@ -0,0 +1,634 @@
+		name: "Input I1 Sensor reporting", key: "inputI1SensorReporting", type: "range",
+		parameterNumber: 140, size: 2, defaultValue: 5,
+		range: "1..10000",
+		description: "Input I1 Sensor reporting If analogue sensor is connected, module reports measured value on change defined by this parameter. " +
```
I'd remove the repetition of the preference name

KKlimczukS replied:
Fixed (that parameter was removed)

KKlimczukS committed:
Fixes

KKlimczukS committed:
Fixes

KKlimczukS committed:
Fixes and workarounds, added two missed device preferences.

KKlimczukS committed:
Fixes

KKlimczukS committed:
Removed "boolRange" preferences boilerplate code

KKlimczukS commented:
> There are no "boolRange" preferences, but a lot of boilerplate code for dealing with them. It should be removed. I'd also like if you could take it out of the walli DTHs since the same is true for them. In general, I'd like the boilerplate to be pared down to the actually represented types of preferences.

@greens 
You are right.  Fixed. The boilerplate code was removed.
It was also fixed in Fibaro Walli DTHs:
https://github.com/SmartThingsCommunity/SmartThingsPublic/pull/34851

greens commented on devicetypes/qubino/child-temperature-sensor.src/child-temperature-sensor.groovy:
```diff
@@ -0,0 +1,53 @@
+ */
+metadata {
+	definition(name: "Child Temperature Sensor", namespace: "qubino", author: "SmartThings", mnmn: "SmartThings", vid: "generic-temperature-measurement", ocfDeviceType: "oic.d.thermostat") {
+		capability "Actuator"
```
this isn't actually an actuator, since it doesn't _do_ anything, right?

KKlimczukS replied:
Good point, that capability shouldn't be there. Fixed.

KKlimczukS committed:
Fixes, corrected few options' names  (for consistency sake)
